tools/benchmark/macros/src/lib.rs

```diff
@@ -16,64 +16,98 @@
     )
 )]
 
-// If no features or all features are present, define empty macros
+/// Re-export tools so that macros can access it without needing a dependency.
+#[cfg(feature = "benchmark_memory")]
+pub use dhat;
+#[cfg(all(
+    feature = "rust_global_allocator",
+    any(target_os = "linux", target_os = "macos", target_arch = "wasm32")
+))]
+pub use dlmalloc;
+
+#[macro_export]
+macro_rules! println {
+    ($lit:literal $(, $arg:expr)* $(,)?) => {{
+        /// For binary size benchmarks we don't want to include `std::fmt::Write` machinery,
+        /// which `println!` pulls in, but we do want to actually evaluate the arguments.
+        #[cfg(not(debug_assertions))]
+        {
+            struct Sink;
+            impl std::fmt::Write for Sink {
+                fn write_str(&mut self, s: &str) -> Result<(), std::fmt::Error> {
+                    std::hint::black_box(s);
+                    Ok(())
+                }
+            }
+
+            $(
+                let _infallible = writeable::Writeable::write_to(&$arg, &mut Sink);
+            )*
+        }
+        #[cfg(debug_assertions)]
+        {
+            std::println!($lit, $($arg,)*);
+        }
+    }};
+}
+
+// If no features or all features are present, do nothing
 #[cfg(any(
     not(any(feature = "benchmark_memory", feature = "rust_global_allocator")),
     all(feature = "benchmark_memory", feature = "rust_global_allocator"),
 ))]
-
-mod default {
-    #[macro_export]
-    macro_rules! static_setup {
-        () => {};
-    }
-    #[macro_export]
-    macro_rules! main_setup {
-        () => {};
-    }
+#[macro_export]
+macro_rules! instrument {
+    () => {
+        const _: () = {
+            #[no_mangle]
+            fn main(_argc: isize, _argv: *const *const u8) -> isize {
```

Comment:
suggestion: we can also just call it `actual_main` and give it a `link_name = main` attribute

Reply:
I tried this, it doesn't seem to work.